### PR TITLE
tikv: implement fallback from async commit or 1PC

### PR DESCRIPTION
**Pull request by sticnarf**
### What problem does this PR solve?

Issue Number: close https://github.com/tikv/sig-transaction/issues/64

### What is changed and how it works?

When TiDB receives `CommitTsTooLarge` error during prewrite, we will fallback from async commit or 1PC.

To recover from this error, we unset async commit or 1PC related fields in the prewrite request and resend the request. At the same time, if the batch is not the primary batch, we send another prewrite request which contains only the primary lock. This request will overwrite the orginal primary lock, making the transaction not an async-commit or 1PC transaction.

Correctness:

Because we fallback to the traditional 2PC, then the most important thing we must guarantee is that all locks are prewritten before we start the commit phase. Possible interference comes from another reader, who reads a lock of this transaction and starts to resolve locks.

- Case 1: If the primary lock does not exist when doing `CheckTxnStatus`, a rollback will be written to prevent prewriting the primary lock. The transaction will fail as expected.

- Case 2: If `CheckTxnStatus` reads the async-commit lock, it will start to check all secondary locks. 
  - 2.1 If `CheckSecondaryLocks` rolls back some locks, at least one prewrite request will fail because of the rollback record. The transaction will fail as expected.
  - 2.2 It is impossible that `CheckSecondaryLocks` succeeds for all locks because at least one lock is not using async commit. Then the resolve lock process will fail.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

- Part of the async commit feature.

### CODE REVIEW

File: store/tikv/async_commit_test.go [hunk 1]
```diff
@@ -395,6 +395,34 @@ func (s *testAsyncCommitSuite) TestAsyncCommitExternalConsistency(c *C) {
 	c.Assert(commitTS2, Less, commitTS1)
 }
 
+func (s *testAsyncCommitSuite) TestAsyncCommitExceedingMaxCommitTS(c *C) {
```
Comment (cfzjywxk) [inline store/tikv/async_commit_test.go:398]: Do we need to trigger the failpoint `commitTsTooLarge` here?

File: store/tikv/prewrite.go [hunk 2]
```diff
@@ -238,3 +275,25 @@ func (c *twoPhaseCommitter) prewriteMutations(bo *Backoffer, mutations Committer
 	// `doActionOnMutations` will unset `useOnePC` if the mutations is splitted into multiple batches.
 	return c.doActionOnMutations(bo, actionPrewrite{}, mutations)
 }
+
+// Reprewrite the primary lock to fallback from async commit or 1PC
+func (c *twoPhaseCommitter) reprewritePrimary(bo *Backoffer) {
+	c.reprewrite.Add(1)
```
Comment (cfzjywxk) [inline store/tikv/prewrite.go:281]: Seems every time we enter the `reprewritePrimary` we will increase the wait value?

File: store/tikv/lock_resolver.go [hunk 3]
```diff
@@ -362,6 +362,11 @@ func (lr *LockResolver) resolveLocks(bo *Backoffer, callerStartTS uint64, locks
 
 			if status.primaryLock != nil && status.primaryLock.UseAsyncCommit && !exists {
 				err = lr.resolveLockAsync(bo, l, status)
+				// Treat nonAsyncCommitLock error as a soft error. It happens when a transaction
```
Comment (cfzjywxk) [inline store/tikv/lock_resolver.go:365]: Some integration test cases are needed for the conflict situation for `resolveAsync` and `fallback and reprewrite`, maybe we could do this in `TiCASE`?

File: store/tikv/prewrite.go (hunk 2 above)
Comment (sticnarf) [reply store/tikv/prewrite.go:281]: Ah yes. It's a mistake. But it's strange that the corresponding unit test didn't fail.

Comment (sticnarf) [reply store/tikv/prewrite.go:281]: I see. I miss the test which fails several secondary prewrite.

File: store/tikv/async_commit_test.go (hunk 1 above)
Comment (sticnarf) [reply store/tikv/async_commit_test.go:398]: We needn't. After setting `SafeWindow` and `allowedClockDrift` to zero, it's sure to get a CommitTsTooLarge error.

**Comment by nrc** [comment]
What's the status with this PR? Are we waiting on landing the tests or does it just need re-review?

**Comment by sticnarf** [comment]
> What's the status with this PR? Are we waiting on landing the tests or does it just need re-review?

I'm working on the integration test. This PR can be reviewed at the same time.

File: store/tikv/lock_resolver.go (hunk 3 above)
Comment (sticnarf) [reply store/tikv/lock_resolver.go:366]: I've created PR https://github.com/pingcap/automated-tests/pull/535 to improve TiCase `TIDB_ASYNC_COMMIT_RESOLVE` to support testing this feature.

**Comment by sticnarf** [comment]
@cfzjywxk @youjiali1995 @MyonKeminta PTAL when you're free

**Comment by sticnarf** [comment]
cc @leiysky @marsishandsome 

This PR contains some changes to ResolveLock: treat nonAsyncCommitLock error as a soft error.

File: store/tikv/prewrite.go [hunk 4]
```diff
@@ -240,3 +283,25 @@ func (c *twoPhaseCommitter) prewriteMutations(bo *Backoffer, mutations Committer
...
+		// But it is not a big problem because fallback happens rarely and the number of mutations are
+		// small in async commit or 1PC.
+		var mutation CommitterMutations
+		for i := 0; i < c.mutations.Len(); i++ {
+			if bytes.Equal(c.primary(), c.mutations.GetKey(i)) {
+				mutation = c.mutations.Slice(i, i+1)
+				break
+			}
+		}
+		c.reprewrite.err = errors.Trace(c.prewriteMutations(bo, mutation))
```
Comment (cfzjywxk) [inline store/tikv/prewrite.go:304]: Should we catch the panic error in this goroutine as it's may block the `execution` path?

File: store/tikv/lock_resolver.go [hunk 5]
```diff
@@ -363,6 +363,11 @@ func (lr *LockResolver) resolveLocks(bo *Backoffer, callerStartTS uint64, locks
...
+				// just fallbacks from async commit, so a simple retry will work.
+				if _, ok := err.(*nonAsyncCommitLock); ok {
+					msBeforeTxnExpired.update(0)
```
Comment (cfzjywxk) [inline store/tikv/lock_resolver.go:369]: It seems to be duplicated with `L377` below?

File: store/tikv/prewrite.go (hunk 4 above)
Comment (sticnarf) [reply store/tikv/prewrite.go:304]: Is it also the reason that there is a `defer cancel` in https://github.com/pingcap/tidb/blob/051285a572/store/tikv/2pc.go#L1661? Then, maybe i would better not remove that cancel?

Comment (cfzjywxk) [reply store/tikv/prewrite.go:313]: I guess so seems we've no such test cases yet.

Comment (cfzjywxk) [reply store/tikv/prewrite.go:313]: I guess so, seems we've no such test cases yet.

Comment (sticnarf) [reply store/tikv/prewrite.go:313]: Panics are not expected to happen so there is no such a test I think. Anyway, I can add one for it.

File: store/tikv/lock_resolver.go (hunk 5 above)
Comment (sticnarf) [reply store/tikv/lock_resolver.go:371]: The error should be ignore. Fixed.

File: store/tikv/lock_resolver.go [hunk 6]
```diff
@@ -363,6 +363,14 @@ func (lr *LockResolver) resolveLocks(bo *Backoffer, callerStartTS uint64, locks
 
 			if status.primaryLock != nil && status.primaryLock.UseAsyncCommit && !exists {
 				err = lr.resolveLockAsync(bo, l, status)
+				// Treat nonAsyncCommitLock error as a soft error. It happens when a transaction
+				// just fallbacks from async commit, so a simple retry will work.
+				if _, ok := err.(*nonAsyncCommitLock); ok {
```
Comment (youjiali1995) [inline store/tikv/lock_resolver.go:368]: Seems it never be true.
```go
	var errs []string
	for range regions {
		err1 := <-errChan
		if err1 != nil {
			errs = append(errs, err1.Error())
		}
	}

	if len(errs) > 0 {
		return nil, errors.Errorf("async commit recovery (sending CheckSecondaryLocks) finished with errors: %v", errs)
	}
```

File: store/tikv/prewrite.go [hunk 7]
```diff
@@ -204,6 +225,29 @@ func (action actionPrewrite) handleSingleBatch(c *twoPhaseCommitter, bo *Backoff
...
+				// Rewrite the request
+				prewriteReq := req.Req.(*pb.PrewriteRequest)
+				prewriteReq.UseAsyncCommit = false
+				prewriteReq.TryOnePc = false
+				prewriteReq.Secondaries = nil
+				prewriteReq.MaxCommitTs = 0
+				prewriteReq.LockTtl = newLockTTL
+				// if this is not a primary batch, we also need to overwrite the primary lock
+				if !batch.isPrimary {
+					c.reprewritePrimary(bo)
```
Comment (youjiali1995) [inline store/tikv/prewrite.go:247]: Seems it's not necessary? Lock resolve can handle all cases now.

Comment (sticnarf) [reply store/tikv/prewrite.go:247]: Sorry I don't understand it. If the primary lock is not changed to non-async-commit, then the primary lock cannot be cleaned by `CheckTxnStatus`.

File: store/tikv/lock_resolver.go (hunk 6 above)
Comment (sticnarf) [reply store/tikv/lock_resolver.go:368]: I should also add test to check resolve lock shouldn't return an error in such a scenario...

**Comment by sticnarf** [comment]
I realize I make the solution too complex. Modifying the process of prewrite looks not so solid.

The better solution is to let TiKV just accept prewrites with larger commit ts, but return 0 as min_commit_ts. Then, the committer can do traditional 2pc.
The primary lock is rewritten during resolve lock. After re-prewrite, the lock resolver checks the primary lock again and does normal lock resolving.

I will provide a new PR later today.

**Pull request closed without merging**